### DISCUSSION TITLE
Should we expect giving AST is clean by default?

### DISCUSSION THREAD

**Discussion by hyf0** [web-infra-dev/rspack#357] [General]
Maybe we should make sure if a folder or a visitor mutated `SyntaxContext`, then It should have the responsibility to clear the mark then return the ast directly.

_Originally posted by @h-a-n-a in https://github.com/speedy-js/rspack/pull/355#discussion_r887549115_

**Comment by hyf0** [comment]
> > Some folders may not obey these rules. For some AST nodes, after`react plugin` mutate the AST, the `SyntaxContext` may be wrong.
> 
> Even if we cannot sure whether the internal passes in swc obey the rule of cleaning context like `HygieneRemover` did, we can make sure our passes would obey these rules, otherwise it may results to a big overhead. by @h-a-n-a

**Comment by hyf0** [comment]
When you are free, any opinions @kdy1?

**Reply by kdy1** [reply]
Are you using `SyntaxContext` for metadata?

I'm not sure what you are doing with `SyntaxContext`

**Reply by kdy1** [reply]
If it's a metadata used by only single pass, cleaning the context after using it seems like a good idea

**Reply by hyf0** [reply]
Our react plugin will [use Resolver and ReactVisitor](https://github.com/speedy-js/rspack/blob/ac89459b72933ee07b9ae222d458bba816adf3dd/crates/rspack_plugin_react/src/lib.rs#L87-L93) to transform the giving AST and inject some HMR runtime. I think the resolved AST is dirty after injecting, right?

And there are other places, we want to check whether an identifier is in the root or global scope. I need to first clean up the dirty AST(using ClearMark), then resolve it with Resolver to get a correct result.

So, there are two times calls of ClearMark, that's because we think the giving AST might be dirty.

If we promise the giving AST is clean by default, we may reduce the calling times of ClearMark.

I now prefer the "clean by default" way, then there are duties for developers to keep promises. We would like to hear your opinion about this, does the same problem happen when you are developing SWC?

--- edited

Thanks. Saw your reply

**Reply by kdy1** [reply]
> Our react plugin will [use Resolver and ReactVisitor](https://github.com/speedy-js/rspack/blob/ac89459b72933ee07b9ae222d458bba816adf3dd/crates/rspack_plugin_react/src/lib.rs#L87-L93) to transform the giving AST and inject some HMR runtime. I think the resolved AST is dirty after injecting, right?

I think so.

> And there are other places, we want to check whether an identifier is in the root or global scope. I need to first clean up the dirty AST(using ClearMark), then resolve it with Resolver to get a correct result.

If you are going to use resolver, yes. You have to run `ClearMark`.
Most part of swc does not modify the syntax context of identifiers because of this reason.
